lib/src/repository/git_local_repository.dart

```diff
@@ -0,0 +1,159 @@
+      return await repo.detectDefaultBranch();
+    } on GitToolException catch (e, st) {
+      log.warning(e.toString(), st);
+      return null;
+    } finally {
+      await repo?.delete();
+    }
+  });
+}
+
+class GitLocalRepository {
+  final String localPath;
+  final String origin;
+
+  GitLocalRepository({
+    required this.localPath,
+    required this.origin,
+  });
+
+  static Future<GitLocalRepository> createLocalRepository(String origin) async {
+    final tempDir = await Directory.systemTemp.createTemp('git-repo');
+    final repo = GitLocalRepository(
+      localPath: tempDir.path,
+      origin: origin,
+    );
+    try {
+      await repo._init();
+      return repo;
+    } on Exception catch (_) {
+      await tempDir.delete(recursive: true);
+      rethrow;
+    }
+  }
+
+  Future<PanaProcessResult> _runGit(
+    List<String> args, {
+    GitToolException Function(PanaProcessResult pr)? createException,
+  }) async {
+    final pr = await runProc(
+      ['git', ...args],
+      workingDirectory: localPath,
+    );
+    if (pr.exitCode != 0) {
+      final ex = createException == null
+          ? GitToolException.failedToRun(args.join(' '), pr)
+          : createException(pr);
+      throw ex;
+    }
+    return pr;
+  }
+
+  Future<PanaProcessResult> _runGitWithRetry(
+    List<String> args, {
+    int maxAttempts = 3,
+    GitToolException Function(PanaProcessResult pr)? createException,
+  }) async {
+    return await retry(
+      () => _runGit(
+        args,
+        createException: createException,
+      ),
+      maxAttempts: maxAttempts,
+      retryIf: (e) => e is GitToolException,
+    );
+  }
+
+  Future<void> _init() async {
+    await _runGit(['init']);
+    await _runGit(['remote', 'add', 'origin', origin]);
+  }
+
+  /// Detects the default branch name by querying the origin and
+  /// matching `HEAD branch` in the output.
+  ///
+  /// Throws [GitToolException] if the git command fails, if the
+  /// branch name is missing, or if the default branch name has
+  /// unexpected pattern.
+  Future<String> detectDefaultBranch() async {
+    final pr = await _runGitWithRetry(['remote', 'show', 'origin']);
+    final output = pr.stdout.toString();
+    final lines = output.split('\n');
+    for (final line in lines) {
+      final parts = line.trim().split(':');
+      if (parts.length == 2 && parts[0].trim() == 'HEAD branch') {
+        final branch = parts[1].trim();
+        if (_acceptedBranchNameRegExp.matchAsPrefix(branch) != null) {
+          return branch;
+        } else {
+          throw GitToolException('Could not accept branch name: `$branch`.');
+        }
+      }
+    }
+    throw GitToolException('Could not find HEAD branch.', output);
+  }
+
+  /// Return the String content of the file in [branch] and [path].
+  ///
+  /// Throws [GitToolException] if the git command fails.
+  Future<String> showStringContent(String branch, String path) async {
```

Comment:
You should wrap `branch` and `path` some sanity checks...
 * Check they are not empty
 * Check that `branch` doesn't contain `/`
 * Sanity checks for special characters?

Throw `ArgumentError` if something is wrong, and document any non-obvious constraints.
A branch like `foo/bar` could certainly cause bugs to happen.

Reply:
The public interface for `GitLocalRepository` does not enforce any patterns, or tell the users what strings can safely be used as `branch`.

It would be safer if we just limit branch names to something sane.

You have a good point that all callers of `GitLocalRepository.showStringContent` in the current code, does have some patterns protecting against branch names that contain `:` or `/`. But the method is public, and I assume we made an abstraction called `GitLocalRepository` because we imagine more things will use this in the future.

So adding an extra check inside `GitLocalRepository.showStringContent` protects us against such mistakes in the future.